compiler/src/iree/compiler/Dialect/Util/Transforms/OptimizeIntArithmetic.cpp

```diff
@@ -186,6 +189,132 @@ struct ConvertUnsignedI64IndexCastProducerToIndex
   DataFlowSolver &solver;
 };
 
+//===----------------------------------------------------------------------===//
+// Index -> int32 assumption narrowing
+// If we're narrowing `index` values to `i32`, a `util.assume.int` on `index`
+// introduces unnecessary zero-extensions and truncations to/from `index`
+// when introducing assumptions.
+//===----------------------------------------------------------------------===//
+struct RemoveIndexCastForAssumeOfI32
+    : public OpRewritePattern<Util::AssumeIntOp> {
+  RemoveIndexCastForAssumeOfI32(MLIRContext *context, DataFlowSolver &solver)
+      : OpRewritePattern(context), solver(solver) {}
+
+  LogicalResult matchAndRewrite(Util::AssumeIntOp op,
+                                PatternRewriter &rewriter) const override {
+    llvm::SmallBitVector needNarrowing(op.getNumOperands(), false);
+    for (auto [idx, arg] : llvm::enumerate(op.getOperands())) {
+      if (!arg.getType().isIndex())
+        continue;
+      auto castOp = arg.getDefiningOp<arith::IndexCastUIOp>();
+      if (!castOp)
+        continue;
+      Value castIn = castOp.getIn();
+      Type intType = castIn.getType();
+      if (intType.getIntOrFloatBitWidth() > 32)
+        continue;
+
+      needNarrowing[idx] = true;
+    }
+    if (needNarrowing.none())
+      return failure();
+
+    SmallVector<Value> newArgs;
+    newArgs.reserve(op.getNumOperands());
+    for (auto [idx, arg] : llvm::enumerate(op.getOperands())) {
+      if (!needNarrowing[idx]) {
+        newArgs.push_back(arg);
+        continue;
+      }
+      newArgs.push_back(arg.getDefiningOp<arith::IndexCastUIOp>().getIn());
+    }
+    ArrayRef<Attribute> assumptions = op.getAssumptions().getValue();
+    // Ugly hack to avoid recreating the assumptions array.
+    ArrayRef<ArrayAttr> castAssumptions = ArrayRef(
+        static_cast<const ArrayAttr *>(assumptions.data()), assumptions.size());
+    auto newOp = rewriter.create<Util::AssumeIntOp>(op.getLoc(), newArgs,
+                                                    castAssumptions);
```

Comment:
Can you not just do `rewriter.create<Util::AssumeIntOp>(op.getLoc(), newArgs, op.getAssumptions());`

If not we can either just clone with new operands or add a new builder.

Reply:
There's a builder missing - and no really room to insert one

Reply:
wdym there is no room to add a new builder? Can't we add a builder that take `ArrayRef<Value>` and `ArrayAttr` as operands?

Reply:
... Yeah, might not be a bad idea, could add one.

Reply:
@qedawkins Done - turns out I just needed an annotation in the tablegen to get the builder I wanted

Reply:
Nice, thanks! Sorry somehow missed the ping here and delayed the approve :(